### DISCUSSION TITLE
Container Migration/MEC Application Migration within two AdvantEDGE Platforms

### DISCUSSION THREAD

**Discussion by itnor4** [InterDigitalInc/AdvantEDGE#321] [Q&A] [answered]
Hi,

I have few questions about the migration within AdvantEDGE platforms deployed on two different machines. Is it possible to create a scenario with two Edge hosts (i.e. having the Edge hosts on two different machines)? If yes, how can I configure them in the AdvantEDGE platform. Also, with the UE mobility is it possible to do the application migration within AdvantEDGE platforms?

Thank you in advance!

**Comment by roymx** [comment, marked as answer]
Hello @itnor4 ,

Thanks for reaching out.

Yes, it is possible to deploy AdvantEDGE on a Kubernetes cluster composed of 1, 2 or more nodes.
To add nodes to your AdvantEDGE cluster, follow the Runtime Setup, details are [here](https://interdigitalinc.github.io/AdvantEDGE/docs/setup/env-runtime/#step-5---optionally-add-worker-nodes-to-k8s-cluster-details) 

With regards to "Application Migration": two solutions are currently supported
- [ETSI MEC021 Application Mobility service](https://interdigitalinc.github.io/AdvantEDGE/docs/overview/edge-services/ams/) (a standardized solution)
- [Application State Transfer](https://interdigitalinc.github.io/AdvantEDGE/docs/overview/edge-services/asts/) (an AdvantEDGE specific solution)

Both solutions allow to transfer an application state from a _source_ to a _target_ edge application; none of the solution handles automatic instantiation of the _target application instance_
Creation of the _target application instance_ is not integrated in these solutions but can be achieved separately using AdvantEDGE  [Process Lifecycle API](https://interdigitalinc.github.io/AdvantEDGE/docs/overview/features/process-lifecycle/) if required.

**Comment by itnor4** [comment]
Hi @roymx,

Two more things: For the additional node, that is able to reach out to the meep-docker-registry, how it can be added as the MEP in the scenario/ sandbox or the scenario doesn't include the additional node information.

As per the above mentioned solutions, we have tried out the Application state transfer API, particularly this solution helps to achieve the migration with the application and as per the UE movement. The migration creates the connectivity without the application context (In our case, the application is video. After the migration the video starts running again from the start instead of the last state. )

Thank you so much, I look forward to hearing from you.

**Reply by roymx** [reply]
Hello @itnor4 ,

**Regarding the first thing**, you are mixing up two layers of the AdvantEDGE platform. `meep-docker-registry` is a _platform level_ service, it should not be accessed directly by the _scenario level_ at scenario runtime. As such, if your experiments require to create a new edge server at scenario runtime, we recommend that you do it via the AdvantEDGE API; more specifically, look at `Compute events` in the [Events subsystem](https://interdigitalinc.github.io/AdvantEDGE/docs/overview/features/events/) and previously mentioned [Process Lifecycle API](https://interdigitalinc.github.io/AdvantEDGE/docs/overview/features/process-lifecycle/)-- doing it this way will not only create the new Pod/Container, but also update the necessary internal AdvantEDGE structures.

Additionally, referring to ETSI MEC architecture, you can see that orchestration is not performed by the MEP but by another entity called the MEO (e.g. MEC Orchestrator); AdvantEDGE platform offers an API that allows a user or a user application to act like the orchestrator either via the scenario definition or by generating events.

**With regards to the second thing,** -- if you transfer the _current video time_ in the user state, then it should be possible to start the video at the target at that specific time and avoid replaying the video from the start.
